src/backend/distributed/utils/node_metadata.c

```diff
@@ -145,6 +148,67 @@ master_initialize_node_metadata(PG_FUNCTION_ARGS)
 
 
 /*
+ * get_shard_id_for_distribution_column function takes a distributed table name and a
+ * distribution value then returns shard id of the shard which belongs to given table and
+ * contains given value. This function only works for hash distributed tables.
+ */
+Datum
+get_shard_id_for_distribution_column(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	Datum distributionValue = PG_GETARG_DATUM(1);
+
+	Var *distributionColumn = NULL;
+	char distributionMethod = 0;
+	Oid expectedElementType = InvalidOid;
+	Oid inputElementType = InvalidOid;
+	DistTableCacheEntry *cacheEntry = NULL;
+	int shardCount = 0;
+	ShardInterval **shardIntervalArray = NULL;
+	FmgrInfo *hashFunction = NULL;
+	FmgrInfo *compareFunction = NULL;
+	bool useBinarySearch = true;
+	ShardInterval *shardInterval = NULL;
+
+	if (!IsDistributedTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+						errmsg("relation is not distributed")));
+	}
+
+	distributionMethod = PartitionMethod(relationId);
+	if (distributionMethod != DISTRIBUTE_BY_HASH)
```

Comment:
Agree with that however from discussion [here](https://github.com/citusdata/citus/issues/872#issuecomment-266018791) @ozgune says;

>> 1) Should this function also support append and range distributed tables?
>
> We think it doesn't needs to. In particular, @samay-sharma noted that this function is more relevant for multi-tenant deployments and if the function returned an array, it would introduce confusion.